### PR TITLE
feat(types): support ns timestamp

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://raw.githubusercontent.com/risingwavelabs/risingwave/17af8a747593ebdbfa826691daf75bdab7d14fa0/.github/contributor-license-agreement.txt).

## What's changed and what's your intention?
our timestamp only support ms, But our internal implementation is capable of ns, it's just that we automatically convert it to ms when we read it in, and this pr can read ns to support ns.

## Checklist

- [ ] I have written necessary rustdoc comments.
- [ ]  I have added necessary unit tests and integration tests.
- [ ]  I have added test labels as necessary. 
- [ ]  I have added fuzzing tests or opened an issue to track them. 
- [ ]  My PR contains breaking changes. 
- [ ]  My PR changes performance-critical code, so I will run (micro) benchmarks and present the results. 
- [ ]  My PR contains critical fixes that are necessary to be merged into the latest release. 

## Documentation

- [ ]  My PR needs documentation updates. 

<details>
<summary><b>Release note</b></summary>

</details>

### CODE REVIEW

File: src/common/src/types/datetime.rs [hunk 1]
```diff
@@ -495,18 +491,24 @@ impl Timestamp {
...
-        output
-            .write(&(self.0.and_utc().timestamp_micros()).to_be_bytes())
-            .map_err(Into::into)
+        let timestamp_size = output
+            .write(&(self.0.and_utc().timestamp()).to_be_bytes())
+            .map_err(Into::<ArrayError>::into)?;
+        let timestamp_subsec_nanos_size = output
+            .write(&(self.0.and_utc().timestamp_subsec_nanos()).to_be_bytes())
+            .map_err(Into::<ArrayError>::into)?;
+        Ok(timestamp_subsec_nanos_size + timestamp_size)
```
Comment (fuyufjh) [inline src/common/src/types/datetime.rs:511]: This will break the compatibility of Timestamp stored in protobuf. Not sure whether it's used somewhere in our catalog.

Comment (fuyufjh) [reply src/common/src/types/datetime.rs:511]: One of the ways to keep **backward** compatibility is: using a special value such as `-1` as a placeholder indicating the new format. That is

- If read `-1`, parse the subsequent 12 bytes in new format
- If read otherwise, consider it as a timestamp_micros (old format)

**Comment by wcy-fdu** [comment]
I tested whether I could read nanosecond in the parquet file on this branch, and it seemed to work.
image

**Comment by wcy-fdu** [comment]
sorry, mis clicked update branch

**Comment by wcy-fdu** [comment]
cc @xiangjinwu  PTAL❤️

File: src/common/src/types/datetime.rs [hunk 2]
```diff
@@ -168,29 +168,18 @@ impl FromStr for Timestamp {
...
-        } else {
-            let res =
-                speedate::Date::parse_str_rfc3339(s).map_err(|_| ErrorKind::ParseTimestamp)?;
-            Ok(
-                Date::from_ymd_uncheck(res.year as i32, res.month as u32, res.day as u32)
-                    .and_hms_micro_uncheck(0, 0, 0, 0),
-            )
-        }
+        let dt = s
+            .parse::<jiff::civil::DateTime>()
```
Comment (xiangjinwu) [inline src/common/src/types/datetime.rs:172]: Why do we need to introduce yet another dependency in addition to `chrono` and `speedate`?

File: src/common/src/types/datetime.rs [hunk 3]
```diff
@@ -495,18 +491,47 @@ impl Timestamp {
     }
 
     pub fn from_protobuf(cur: &mut Cursor<&[u8]>) -> ArrayResult<Timestamp> {
-        let micros = cur
+        let secs = cur
```
Comment (xiangjinwu) [inline src/common/src/types/datetime.rs:494]: nit: `secs_or_usecs`

File: src/common/src/types/datetime.rs [hunk 4]
```diff
@@ -495,18 +491,47 @@ impl Timestamp {
...
+        value ^ (0b01 << 62)
+    }
+
+    fn has_timestamp_namo_format_state(value: i64) -> bool {
+        let state = (value >> 62) & 0b11;
+        state == 0b10 || state == 0b01
+    }
+
+    fn remove_timestamp_namo_format_state(value: i64) -> i64 {
+        value ^ (0b01 << 62)
```
Comment (xiangjinwu) [inline src/common/src/types/datetime.rs:534]: Encapsulate as a new type:
```
/// document about old and new format, including the meaning of highest 2 bits, and the corresponding accepted ranges.
enum FirstI64 {
  V0{ usecs: i64 }
  V1{ secs: i64 }
}
impl From<i64> for FirstI64 {
  // check and remove the tagging bits
}
impl From<FirstI64> for i64 {
  // add the tagging bits
}
```

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 5]
```diff
@@ -495,7 +495,7 @@ impl HummockVersion {
                             .clone_from(&group_construct.table_ids);
                         self.levels.insert(*compaction_group_id, new_levels);
                         let member_table_ids = if group_construct.version
-                            >= CompatibilityVersion::NoMemberTableIds as _
+                            >= CompatibilityVersion::NoMemberTableIds as i32
```
Comment (xiangjinwu) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:498]: avoid unrelated change

**Review by xiangjinwu — COMMENTED** [review]
* Could you also include the testing SQLs you have tried?
* Is any caller of `to_protobuf`/`from_protobuf` assuming that `Timestamp` would be fixed 64-bit? This assumption is no longer true.

**Comment by xxhZs** [comment]
> * Could you also include the testing SQLs you have tried?

add ci

File: src/common/src/types/datetime.rs (hunk 2 above)
Comment (xxhZs) [reply src/common/src/types/datetime.rs:172]: Because chrono can't be parsed, and speedate is only parsed at the us level

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 5 above)
Comment (xiangjinwu) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs]: It seems we do need this as long as we use `jiff` in `src/common/src/types/datetime.rs`. Not sure why though ... maybe a rustc bug.

Btw it can be `group_construct.version() >= CompatibilityVersion::NoMemberTableIds` rather than `if group_construct.version >= CompatibilityVersion::NoMemberTableIds as i32`

File: e2e_test/batch/types/timestamp_ns.slt.part [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+1  2013-01-01 01:01:01.123456+00:00
+4  2213-01-01 01:01:01.123456+00:00
+2  2012-01-01 01:01:01.123456+00:00
+
+query T
+select v1, to_char(v2, 'YYYY-MM-DD HH24:MI:SS.NS') as formatted_v2 from t1;
+----
+3  0001-01-01 01:01:01.123456789 BC
+5  null
+1  2013-01-01 01:01:01.123456000
```
Comment (xiangjinwu) [inline e2e_test/batch/types/timestamp_ns.slt.part:101]: Did this case pass? Is it supposed to be `2013-01-01 01:01:01.123456789`?

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 5 above)
Comment (xxhZs) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:498]: It doesn't seem to work because the type of CompatibilityVersion::NoMemberTableIds is not i32

Comment (xiangjinwu) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:498]: `group_construct.version()` returns `CompatibilityVersion`, while `group_construct.version` is `i32`

**Comment by xiangjinwu** [comment]
Btw you can fix the e2e test with `rowsort`.

https://www.sqlite.org/sqllogictest/doc/trunk/about.wiki
> The default is "nosort". In nosort mode, the results appear in exactly the order in which they were received from the database engine. The nosort mode should only be used on queries that have an ORDER BY clause or which only have a single row of result, since otherwise the order of results is undefined and might vary from one database engine to another. The "rowsort" mode gathers all output from the database engine then sorts it by rows on the client side.

#6776
> If there's `rowsort`, slt file will contain sorted results.

**Comment by xiangjinwu** [comment]
FYI regarding your latest e2e failure, it is actually a bug in our existing `to_char` implementation:

`1 BC` in PostgreSQL is year `0` in chrono. Although chrono also provides `DateLike::year_ce` method, its `strftime` module does not provide a specifier to use `year_ce` instead of `year`.

Given it is out of scope of this PR, let's just replace the case with one without `BC`.

File: src/common/src/types/datetime.rs [hunk 7]
```diff
@@ -168,29 +168,18 @@ impl FromStr for Timestamp {
...
-            let res =
-                speedate::Date::parse_str_rfc3339(s).map_err(|_| ErrorKind::ParseTimestamp)?;
-            Ok(
-                Date::from_ymd_uncheck(res.year as i32, res.month as u32, res.day as u32)
-                    .and_hms_micro_uncheck(0, 0, 0, 0),
-            )
-        }
+        let dt = s
+            .parse::<jiff::civil::DateTime>()
+            .map_err(|_| ErrorKind::ParseTimestamp)?;
```
Comment (xiangjinwu) [inline src/common/src/types/datetime.rs:173]: Error message of `ParseTimestamp` shall be updated from _up to **6** digits_ to _up to **9** digits_.

**Review by xiangjinwu — APPROVED** [review]
Mostly looks good to me. Just more to test:
* `distinct` / `group by` / `order by` / `max`
* `extract` with `epoch` / `second` / `milliseconds` / `microseconds `
* `make_timestamp`

**Comment by wcy-fdu** [comment]
Thanks for the efforts. I will doing some testing on sink side later.

**Comment by xiangjinwu** [comment]
> * Is any caller of `to_protobuf`/`from_protobuf` assuming that `Timestamp` would be fixed 64-bit? This assumption is no longer true.

Answering my own question: there is but it is only used as a hint - so okay to be inaccurate.

`to_protobuf` is used here. The `size_of::<T>` may result in an inaccurate estimate for buffer capacity.

https://github.com/risingwavelabs/risingwave/blob/05397f91957f9ca9ed7c893ef6b667faf1bae4a0/src/common/src/array/primitive_array.rs#L232-L236

`from_protobuf` is used here. Movement of `Cursor` makes no assumption on item sizes.

https://github.com/risingwavelabs/risingwave/blob/05397f91957f9ca9ed7c893ef6b667faf1bae4a0/src/common/src/array/proto_reader.rs#L69-L76

File: src/expr/impl/src/scalar/extract.rs [hunk 8]
```diff
@@ -42,17 +42,18 @@ fn extract_from_datelike(date: impl Datelike, unit: Unit) -> Decimal {
 
 /// Extract field from `Timelike`.
 fn extract_from_timelike(time: impl Timelike, unit: Unit) -> Decimal {
-    let usecs = || time.second() as u64 * 1_000_000 + (time.nanosecond() / 1000) as u64;
+    let usecs = || time.second() as u64 * 1_000_000_000 + time.nanosecond() as u64;
```
Comment (xiangjinwu) [inline src/expr/impl/src/scalar/extract.rs:45]:
```suggestion
    let nanos = || time.second() as u64 * 1_000_000_000 + time.nanosecond() as u64;
```

**Comment by pjpringle** [comment]
Nice useful feature for trading use cases
